components/style/data.rs

```diff
@@ -183,65 +185,124 @@ impl From<RestyleHint> for StoredRestyleHint {
+
+    pub fn destroy(&mut self) {
+        self.destroyed = true;
+        debug_assert!(self.is_none());
+    }
+
+    pub fn ensure<F: FnOnce() -> Snapshot>(&mut self, create: F) -> &mut Snapshot {
+        debug_assert!(thread_state::get().is_layout());
+        if self.is_none() {
+            self.snapshot = Some(create());
+            self.destroyed = false;
+        }
+
+        self.snapshot.as_mut().unwrap()
+    }
+}
+
+impl Deref for SnapshotOption {
+    type Target = Option<Snapshot>;
+    fn deref(&self) -> &Option<Snapshot> {
+        if self.destroyed {
+            &NO_SNAPSHOT
+        } else {
+            &self.snapshot
+        }
+    }
 }
 
 /// Transient data used by the restyle algorithm. This structure is instantiated
 /// either before or during restyle traversal, and is cleared at the end of node
 /// processing.
 #[derive(Debug)]
 pub struct RestyleData {
-    pub styles: RestyleDataStyles,
+    pub styles: ElementStyles,
     pub hint: StoredRestyleHint,
+    pub recascade: bool,
     pub damage: RestyleDamage,
-    pub snapshot: Option<Snapshot>,
+    pub snapshot: SnapshotOption,
 }
 
 impl RestyleData {
-    fn new(previous: ElementStyles) -> Self {
+    fn new(styles: ElementStyles) -> Self {
         RestyleData {
-            styles: RestyleDataStyles::Previous(previous),
+            styles: styles,
             hint: StoredRestyleHint::default(),
+            recascade: false,
             damage: RestyleDamage::empty(),
-            snapshot: None,
+            snapshot: SnapshotOption::empty(),
         }
     }
 
-    pub fn get_current_styles(&self) -> Option<&ElementStyles> {
-        use self::RestyleDataStyles::*;
-        match self.styles {
-            Previous(_) => None,
-            New(ref x) => Some(x),
+    /// Expands the snapshot (if any) into a restyle hint. Returns true if later siblings
+    /// must be restyled.
+    pub fn expand_snapshot<E: TElement>(&mut self, element: E, stylist: &Stylist) -> bool {
+        if self.snapshot.is_none() {
+            return false;
+        }
+
+        // Compute the hint.
+        let state = element.get_state();
+        let mut hint = stylist.compute_restyle_hint(&element,
+                                                    self.snapshot.as_ref().unwrap(),
+                                                    state);
+
+        // If the hint includes a directive for later siblings, strip it out and
+        // notify the caller to modify the base hint for future siblings.
+        let later_siblings = hint.contains(RESTYLE_LATER_SIBLINGS);
+        if later_siblings {
+            hint.remove(RESTYLE_LATER_SIBLINGS);
         }
+
+        // Insert the hint.
+        self.hint.insert(&hint.into());
+
+        // Destroy the snapshot.
+        self.snapshot.destroy();
+
+        later_siblings
     }
 
-    pub fn current_styles(&self) -> &ElementStyles {
-        self.get_current_styles().unwrap()
+    pub fn has_current_styles(&self) -> bool {
```

Comment:
Probably `has_up_to_date_styles` is a bit clearer? A doc comment is enough, probably.